### ISSUE TITLE
False positive WordPress.WP.PreparedSQL.NotPrepared

### ISSUE THREAD

**Issue by paulschreiber** [WordPress-Coding-Standards/WordPress-Coding-Standards#469]
I'm getting this error:

    Use placeholders and $wpdb->prepare(); found $post_query (WordPress.WP.PreparedSQL.NotPrepared)

Here's the code:
```php
$post_query = "SELECT ID, post_content, post_name, post_title, DATE_FORMAT(post_date, '%%Y/%%m/%%d %%H:%%i') AS post_date
							FROM {$wpdb->posts}
							WHERE post_status = 'publish' and post_type = %s
							ORDER BY post_date DESC";

if ( $limit && is_numeric( $limit ) ) {
	$post_query .= ' LIMIT ' . intval( $limit );
}

if ( $refresh || false === ( $posts = wp_cache_get( 'omniture.posts' ) ) ) {
	$posts = $wpdb->get_results( $wpdb->prepare( $post_query, $post_type ), ARRAY_A );
	wp_cache_set( 'omniture.posts', $posts );
}
```

I am using `$wpdb->prepare()`.

**Comment by JDGrimes** [comment]
The sniff isn't smart enough to follow a variable like `$post_query` back to its source(s). So it doesn't know whether the query in `$post_query` incorporates user-supplied values or not, i.e., it doesn't know if all of the values incorporated into it have really been prepared.

Ideally, there would be a way to refactor your code so that the query string isn't saved to a variable like this.

Alternatively, now that you've manually reviewed the code and can see that everything is properly prepared, you can whitelist this like so:

```php
    $posts = $wpdb->get_results( $wpdb->prepare( $post_query, $post_type ), ARRAY_A ); // WPCS: unprepared SQL OK
```

**Comment by paulschreiber** [comment]
What's a good way to refactor this, given there is a conditional?

**Comment by JDGrimes** [comment]
I'm not sure there is really a "good" way to refactor it. You could duplicate the query, or you could use a ternary condition (if the sniff supported ternaries—it doesn't). I think those are the only options short of a full-scale SQL query builder.

**Comment by westonruter** [comment]
@JDGrimes What do you think about the feasibility of statically looking back to assignments to the `$post_query` variable to come up with a raw concatenated string which can be looked at to check for the presence of formats. It still probably wouldn't be able to catch everything.

**Comment by JDGrimes** [comment]
It is possible, but has its limits. My opinion is that it isn't worth it for what we gain: the ability to write code that is more difficult to manually review. It also opens up more possibility for false negatives.

If we want to build something that complex, we should use the right tool, which in my opinion isn't phpcs. I've toyed with doing it with PHP Parser, but haven't had the time (yet).

If someone else wants to attempt some look-back support, they are welcome to, but I will not be.

**Comment by westonruter** [comment]
Agreed. Whitelisting comment is the solution here.
